src/core/public/workspace/workspaces_client.ts

```diff
@@ -0,0 +1,191 @@
+import { WorkspaceAttribute, WorkspaceFindOptions } from '.';
+
+/**
+ * WorkspacesClientContract as implemented by the {@link WorkspacesClient}
+ *
+ * @public
+ */
+export type WorkspacesClientContract = PublicMethodsOf<WorkspacesClient>;
+
+const join = (...uriComponents: Array<string | undefined>) =>
+  uriComponents
+    .filter((comp): comp is string => Boolean(comp))
+    .map(encodeURIComponent)
+    .join('/');
+
+type IResponse<T> =
+  | {
+      result: T;
+      success: true;
+    }
+  | {
+      success: false;
+      error?: string;
+    };
+
+/**
+ * Workspaces is OpenSearchDashboards's visualize mechanism allowing admins to
+ * organize related features
+ *
+ * @public
+ */
+export class WorkspacesClient {
+  private http: HttpStart;
+  constructor(http: HttpStart) {
+    this.http = http;
+  }
+
+  private getPath(path: Array<string | undefined>): string {
+    return resolveUrl(`${WORKSPACES_API_BASE_URL}/`, join(...path));
+  }
+
+  public async enterWorkspace(id: string): Promise<IResponse<null>> {
+    return {
+      success: false,
+      error: 'Unimplement',
+    };
+  }
+
+  public async exitWorkspace(): Promise<IResponse<null>> {
+    return {
+      success: false,
+      error: 'Unimplement',
+    };
+  }
+
+  public async getCurrentWorkspaceId(): Promise<IResponse<WorkspaceAttribute['id']>> {
+    return {
+      success: false,
+      error: 'Unimplement',
+    };
+  }
+
+  public async getCurrentWorkspace(): Promise<IResponse<WorkspaceAttribute>> {
+    return {
+      success: false,
+      error: 'Unimplement',
+    };
+  }
+
+  /**
+   * Persists an workspace
+   *
+   * @param attributes
+   * @returns
+   */
+  public create = (
+    attributes: Omit<WorkspaceAttribute, 'id'>
+  ): Promise<IResponse<WorkspaceAttribute>> => {
+    if (!attributes) {
+      return Promise.reject(new Error('requires attributes'));
+    }
+
+    const path = this.getPath([]);
+
+    return this.http.fetch(path, {
+      method: 'POST',
+      body: JSON.stringify({
+        attributes,
+      }),
+    });
+  };
+
+  /**
+   * Deletes a workspace
+   *
+   * @param id
+   * @returns
+   */
+  public delete = (id: string): Promise<IResponse<null>> => {
+    if (!id) {
+      return Promise.reject(new Error('requires id'));
+    }
+
+    return this.http.delete(this.getPath([id]), { method: 'DELETE' });
+  };
+
+  /**
+   * Search for workspaces
+   *
+   * @param {object} [options={}]
+   * @property {string} options.search
+   * @property {string} options.search_fields - see OpenSearch Simple Query String
+   *                                        Query field argument for more information
+   * @property {integer} [options.page=1]
+   * @property {integer} [options.per_page=20]
+   * @property {array} options.fields
+   * @returns A find result with workspaces matching the specified search.
+   */
+  public list = (
+    options?: WorkspaceFindOptions
+  ): Promise<
+    IResponse<
+      WorkspaceAttribute & {
```

Comment:
@SuZhou-Joe  There should respond a array of WorkspaceAttribute instead of single?

Reply:
It will be more extensible that we claim the response as promise  and it may require some API calls when we need to fetch the id.

Reply:
You are right, let me modify this.

Reply:
https://github.com/ruanyl/OpenSearch-Dashboards/pull/7/files#diff-78985f31a7d8676a8244eb638e8090f4597042e979465020e880e8e91e03dd07